# Вложенные структуры данных

Ноутбук основан на материалах курса [Python как иностранный](https://online.hse.ru/course/view.php?id=5261)

Дополнительно нужно прорешать рекомендованные задачи из [яндекс-контеста](https://github.com/hse-econ-data-science/dap_2022-23/tree/main/sem06_dict).



Мы с вами говорили про все основные типы данных в языке Python. Давайте вспомним их.

| Как мы их называем | Как они называютcя по-английски | Как они называются в Python | 
| :- | :- | :- |
| **Естественные типы данных** | 
| целое число | integer | `int` |
| вещественное число | floating-point number | `float` |
| логическая переменная | boolean / logical | `bool` |
| **Упорядоченные типы данных (последовательности)** |
| строка | string | `str` |
| список | list | `lst` |
| кортеж | tuple | `tuple` |
| **Неупорядоченные типы данных (коллекции)** |
| множество | set | `set` |
| словарь | dictionary | `dict` |

Мы также делили данные на неизменяемые и неизменяемые. Например, нужно помнить, что только неизменяемые типы данных могут быть элементами множеств или ключами в словарях. Или то, что методы списков меняют списки, к которым применяются, а методы строк создают новую строку, которою нужно не забыть сохранить в переменную, если мы хотим продолжить работу именно с измененной строкой.

| Неизменяемые типы данных | Изменяемые типы данных |
| --- | --- |
| кортеж | список |
| строки | множество
| числа целые и вещественные | словарь |
| логические переменные

Когда мы говорили об упорядоченных и неупорядоченных данных, мы еще говорили о том, как мы можем обратиться к элементам внутри них.

| Тип данных | Тип структуры данных | Как обращаемся к элементу внутри? |
| --- | --- | --- |
| кортежи | упорядоченный | по индексу |
| списки |  упорядоченный | по индексу |
| строки  | упорядоченный | по индексу |
| множество | неупорядоченный | не можем обратиться к элементу |
| Словари | неупорядоченный | по ключу

Нужно заметить, что словарь иногда относят к упорядоченным видам данных — мы все-таки можем обратиться к элементу, хоть и не по индексу. Но классическое определение, все-таки относит его к неупорядоченным, как тип, с которым не работает индексация по порядоковому номеру элемента.

Теперь, когда мы структурировали наши знания, давайте поговорим о *вложенных структурах данных*. Так мы говорили о возможности хранить внутри кортежа или списка любые данные, и о том, что значением в словаре может быть какой угодно тип данных (в отличие от ключей).

На самом деле такие сложные структуры данных будут часто встречаться в программах. Например, мы уже видели, что хранить информацию в словаре гораздо удобнее чем в двух списках одинаковой длины. А словарь, где значения являются списками, по своей структуре уже будет напомнить таблицу. Более сложные структуры позволят нам эффективнее решать более сложные задачи.

## Списки списков

Как понятно из названия `список списков` — это такая вложенная структура, где элементами списка тоже являются списки. Например, в списке списков удобно хранить одинаково структурированную информацию.

In [ ]:
# Представим, что наш список хранит информацию о студентах: имя, год рождения, образовательная программа
students = [['Ольга Ларина', 2003, 'Политология'], ['Вениамин Ерофеев', 2002, 'Химия']]  

print(students[0])  # Видим, что первый элемент списка, не строка "Ольга Ларина", а список
print(students[1])  # И второй тоже список

# Но мы можем добраться до имени студента с помощью двойной индексации — 
# сначала обратимся к индексу списка, который нам нужен, а затем к индексу элемента внутри этого списка
print(students[0][0])  

# Так как строка тоже поддерживает индексацию, то можем пойти еще глубже и достать первую букву 
# первого элемента (имени) первого списка  
print(students[0][0][0]) 

['Ольга Ларина', 2003, 'Политология']
['Вениамин Ерофеев', 2002, 'Химия']
Ольга Ларина
О


Раньше с помощью метода `.append()` мы добавляли в список строки, числа и другие естественные типы данных. Но с помощью `.append()` в список можно добавить и списки, и кортежи, и множества, и словари (да и создать список со всеми этими типами данных в качестве элементов тоже можно).

Добавим в наш список нового студента:

In [ ]:
students.append(['Евгения Онегина', 2006, 'Журналистика'])
print(students)

[['Ольга Ларина', 2003, 'Политология'], ['Вениамин Ерофеев', 2002, 'Химия'], ['Евгения Онегина', 2006, 'Журналистика']]


Можно добавить список прямо с клавиатуры через конструкцию `input().split()`. Нужно только не забыть потом преобразовать все элементы в нужный тип.

In [ ]:
# Считываем строку, где нужные элементы разделены запятой, и превращаем ее в список с помощью метода `.split()`
# Получившийся список добавляем в список students с помощью `.append()`
students.append(input('Введите через запятую: имя и фамилию студента, год рождения, программу обучения: ').split(','))

# В предыдущих записях год рождения у нас типа int, конвертируем строку в целое число
students[-1][1] = int(students[-1][1])  

print(students)

Введите через запятую: имя и фамилию студента, год рождения, программу обучения: Софья Чацкая,2002,Химия
[['Ольга Ларина', 2003, 'Политология'], ['Вениамин Ерофеев', 2002, 'Химия'], ['Евгения Онегина', 2006, 'Журналистика'], ['Софья Чацкая', 2002, 'Химия']]


Элементы такого списка мы можем перебирать внутри цикла.

In [ ]:
for student in students:                                # В цикле перебираем элемента внутри списка students
    print(f'Работаем со списком: {student}')            # Убеждаемся, что в переменную student попадает список целиком
    print(f'Студент: {student[0]}')                     # Печатаем имя студента — первый элемент каждого списка                                   
    print(f'Возраст: {2020 - student[1]}')              # Высчитываем и печатаем возраст студента в 2020 году, 
                                                        # вычитая год рождения — второй элемент каждого списка
    print(f'Программа обучения: {student[2].upper()}')  # Печатаем программу обучения студента — третий элемент 
                                                        # каждого списка
    print('-'*10)                                       # Печатаем отбивку из 10 дефисов, чтобы в выводе обозначить 
                                                        # конец обработки списка студента

Работаем со списком: ['Ольга Ларина', 2003, 'Политология']
Студент: Ольга Ларина
Возраст: 17
Программа обучения: ПОЛИТОЛОГИЯ
----------
Работаем со списком: ['Вениамин Ерофеев', 2002, 'Химия']
Студент: Вениамин Ерофеев
Возраст: 18
Программа обучения: ХИМИЯ
----------
Работаем со списком: ['Евгения Онегина', 2006, 'Журналистика']
Студент: Евгения Онегина
Возраст: 14
Программа обучения: ЖУРНАЛИСТИКА
----------
Работаем со списком: ['София Чацкая', 2002, 'Физика']
Студент: София Чацкая
Возраст: 18
Программа обучения: ФИЗИКА
----------


Более того, мы можем написать вложенный цикл ``for`` и перебрать элементы внутри вложенных списков. В нашем случае. Звучит сложно, а на самом деле не очень! Давайте сначала переберем элементы внутри списка ``students``, а затем для каждого студента переберем элементы списка, в котором записана информация о нем или о ней.

In [ ]:
for student in students:  # В этом цикле перебираем вложенные списки внутри списка students
    print(f'Вывожу информацию о студенте:')
    for elem in student:  # А в этом цикле для каждого вложенного списка перебираем его элементы
        print(elem)
    print('-'*10)

Вывожу информацию о студенте:
Ольга Ларина
2003
Политология
----------
Вывожу информацию о студенте:
Вениамин Ерофеев
2002
Химия
----------
Вывожу информацию о студенте:
Евгения Онегина
2006
Журналистика
----------
Вывожу информацию о студенте:
София Чацкая
2002
Физика
----------


Можно использовать и цикл ``range()``, чтобы работать не напрямую с элементами, а обращаться к ним по индексам. На самом деле к этому моменту мы знаем уже достаточно много, чтобы находить самые разные комбинации различных инструкций и типов данных.

In [ ]:
for i in range(len(students)):                       # Запускаем цикл по интервалу индексов элемента списка students
    print(f'Вывожу информацию о студенте № {i+1}:')  # Корректируем индекс на 1, чтобы было понятно людям, а не машинам
    for elem in students[i]:                         # А теперь по индексу обращаемся к вложенному списку и 
        print(elem)                                  # перебираем элементы внутри него  
    print('-'*10)

Вывожу информацию о студенте № 1:
Ольга Ларина
2003
Политология
----------
Вывожу информацию о студенте № 2:
Вениамин Ерофеев
2002
Химия
----------
Вывожу информацию о студенте № 3:
Евгения Онегина
2006
Журналистика
----------
Вывожу информацию о студенте № 4:
София Чацкая
2002
Физика
----------


Давайте попробуем решить следующую задачу. Напишем программу, которая будет считывать текущие оценки для нескольких предметов и выводить итоговую оценку за предмет.

1. Создадим пустой список, в котором будем хранить списки оценок по предметам.

2. Сначала мы считаем названия предметов, итоговые оценки за которые будем учитывать при подсчете среднего, и превратим строку в список, разделив их по заяпятой.

3. Потом в цикле для каждого предмета мы считаем оценки, которые мы получили за семестетр.
 * Сначала считаем оценки, введеные через пробел, и превратим их в список.
 * Потом с помощью функции ``map()`` превратим наши оценки-строки в оценки-вещественные числа.
 * Превратим результат работы функции ``map()`` в список (чтобы не только компьютер, но и мы увидели, что там внутри)
 * Добавим список оценок в список ``marks`` с помощью метода `.append()`

4. В цикле переберем наши списки оценок внутри списка ``marks`` и для каждого предмета найдем итоговую оценку по этому предмету — среднее арифметическое.      
  * Будем считать, что наши преподватели очень строгие, поэтому не округляют оценки, а просто отбрасывают дробную часть (так 7.6 станет 7). 
  * Выведем название предмета и итоговую оценку. 

Количество предметов и оценок для каждого предмета может быть произвольным (но как минимум один предмет и одна оценка).

In [ ]:
# Создаем пустой список, будем добавлять сюда списки оценок
marks = []  

# Считываем в список названия предметов
courses = input('Введите через запятую названия предметов, для которых будем считать оценки:\n').split(',')

# Запускаем цикл для всех предметов по индексам
for i in range(len(courses)):  
    # Считывам оценки за предмет в список
    course_marks = input(f'Введите через пробел оценки текущего контроля для предмета "{courses[i]}":\n').split()
    marks.append(list(map(float, course_marks)))  # Преобразуем оценки из строк в вещественные числа и 
                                                  # добавим список с вещественными числами в список marks

# Считали все оценки, теперь посчитаем итоговую оценку для каждого предмета
# Запускаем цикл по индексам для списка списков оценок marks
for i in range(len(marks)):  
    final_mark = int(sum(marks[i])/len(marks[i]))                       # Находим среднюю оценку: делим сумму элементов на 
                                                                        # количество элементов, и отбрасываем дробную часть
    print(f'Итоговая оценка по предмету "{courses[i]}": {final_mark}')  # Печатаем название предмета и итоговую оценку


Введите через запятую названия предметов, для которых будем считать оценки:
сольфеджио,введение в защиту от темных искусств
Введите через пробел оценки текущего контроля для предмета "сольфеджио":
10 8 7 6 5
Введите через пробел оценки текущего контроля для предмета "введение в защиту от темных искусств":
10 8 9 10 2
Итоговая оценка по предмету "сольфеджио": 7
Итоговая оценка по предмету "введение в защиту от темных искусств": 7


Как мы уже говорили, элементами списков могут быть любые из известных нам типов и структур данных в Python. Поэтому, если вдруг вы понимаете, что для решения вашей задачи может пригодиться список множест, то ни в чем себе не отказывайте.

# Словари списков
Мы можем хранить в словаре не только строки или числа, но и списки с кортежами. Например, если мы запоминаем не просто температуру, а температуру днём и ночью, мы можем к каждому ключу (дате) указать список из двух чисел:

In [ ]:
temperature = {
    '1 января': [-5, -12],
    '2 января': [-4, -8]
}

# Выведем весь словарь температур — для двух дней:
print(f'Температура по дням (дневная, ночная): {temperature}')

# Выведем список температур — для 1 января:
print(f'Температура 1 января (дневная, ночная): {temperature["1 января"]}')

# Выведем дневную температуру 1 января:
print(f'Температура 1 января (дневная): {temperature["1 января"][0]}')

Температура по дням (дневная, ночная): {'1 января': [-5, -12], '2 января': [-4, -8]}
Температура 1 января (дневная, ночная): [-5, -12]
Температура 1 января (дневная): -5


Когда мы пишем `temperature["1 января"][0]`, мы просим компьютер:
1. найти в словаре температур данные за 1 января (по ключу "1 января");
2. в списке температур за 1 января взять элемент с 0 индексом.

Мы можем перебрать словарь температур в цикле ``for``, и каждый раз получать доступ к списку температур для каждого дня:

In [ ]:
for day in temperature:
    print(f'{day} ночью {temperature[day][1]}°C, днём {temperature[day][0]}°C.')
    print(f'Среднесуточная температура составила {(temperature[day][1]+temperature[day][0])/2}°C.')

1 января ночью -12°C, днём -5°C.
Среднесуточная температура составила -8.5°C.
2 января ночью -8°C, днём -4°C.
Среднесуточная температура составила -6.0°C.


Со списками, хранящимися внутри значений словаря, можно делать то же, что и с обычными списками, например, мы можем переписать верхний пример через функции `sum()` и `len()`:

In [ ]:
for day in temperature:
    print(f'{day} ночью {temperature[day][1]}°C, днём {temperature[day][0]}°C.')
    print(f'Среднесуточная температура составила {sum(temperature[day])/len(temperature[day])}°C.')

1 января ночью -12°C, днём -5°C.
Среднесуточная температура составила -8.5°C.
2 января ночью -8°C, днём -4°C.
Среднесуточная температура составила -6.0°C.


Элементы внутри значения-списка можно перебирать так же, как и элементы вложенного списка, что мы делали в прошлый раз.

Предположим, что каждый день у нас было много замеров температуры, и нас интересует разница между средней температурой замеров и каждым замером, тогда мы можем:
1. посчитать сумму всех замеров;
2. посчитать количество всех замеров;
3. получить среднее;
4. перебрать все дни;
    * для каждого дня перебрать все замеры, и для каждого из них вывести разность.



In [ ]:
temperature = {
    '1 января': [-5, -12, -8],
    '2 января': [-4, -8, -6],
    '3 января': [-2, -4, -5]
}
total = 0   # Сколько всего замеров
amount = 0  # Сумма всех замеров

for day in temperature:                 # Просмотрим все дни
    total += len(temperature[day])      # Добавим количество замеров к total
    amount += sum(temperature[day])     # Добавим сумму температур к amount

average = amount/total                  # Посчитаем среднее
print(f'Средняя температура января: {average}')

for day in temperature:
    print(f'{day}:')
    for t in temperature[day]:
        print(f'{t}. Разница со средним: {average-t}')

Средняя температура января: -6.0
1 января:
-5. Разница со средним: -1.0
-12. Разница со средним: 6.0
-8. Разница со средним: 2.0
2 января:
-4. Разница со средним: -2.0
-8. Разница со средним: 2.0
-6. Разница со средним: 0.0
3 января:
-2. Разница со средним: -4.0
-4. Разница со средним: -2.0
-5. Разница со средним: -1.0


Внутри значений словаря можно хранить не только списки с кортежами, но и множества. 

Посмотрим на словарь, в котором хранятся номера кабинетов и имена студентов, распределенных на экзамен. Каждому ключу в словаре (номеру кабинета) соответствует множество имён студентов, которые будут писать экзамен в этом кабинете:

In [ ]:
exam = { 
    '102': {'Аня', 'Таня'},      
    '225': {'Дима', 'Катя'},
    '404': {},      
    '500': {'Алёна', 'Настя'}
}
print(f'Все кабинеты экзамена: {exam}')

Все кабинеты экзамена: {'102': {'Аня', 'Таня'}, '225': {'Катя', 'Дима'}, '404': {}, '500': {'Настя', 'Алёна'}}


Переберём все кабинеты и выведем множество имён студентов, зарегистрированных на экзамен:

In [ ]:
for room in exam:
    print(f'В кабинете №{room} экзамен сдают: {", ".join(exam[room])}.')

В кабинете №102 экзамен сдают: Аня, Таня.
В кабинете №225 экзамен сдают: Катя, Дима.
В кабинете №404 экзамен сдают: .
В кабинете №500 экзамен сдают: Настя, Алёна.


Добавим проверку, что кто-то сдаёт экзамен в аудитории, проверив размер множества имён:

In [ ]:
for room in exam:
    if len(exam[room]) > 0:
        print(f'В кабинете №{room} экзамен сдают: {", ".join(exam[room])}.')
    else:
        print(f'В кабинете №{room} экзамен не сдают.')

В кабинете №102 экзамен сдают: Аня, Таня.
В кабинете №225 экзамен сдают: Катя, Дима.
В кабинете №404 экзамен не сдают.
В кабинете №500 экзамен сдают: Настя, Алёна.


Вернёмся к нашей библиотеке. Может ли у автора быть больше одной книги? Да, может. Для простоты запомним только список их названий:

In [ ]:
# Создадим словарь, где ключом будет автор, а в качестве значений будет храниться список названий книг
books = { 
    'Норберт Винер': [  # Создадим список и наполним 1 книгой
        'Кибернетика или управление и связь в животном и машине'
    ],
    'Лев Выготский': [  # Создадим список и наполним 2 книгами
        'Мышление и речь',
        'Психология искусства'
    ]
}
print(f'Выготский написал {books["Лев Выготский"]}')

Выготский написал ['Мышление и речь', 'Психология искусства']


Пройдём по авторам (ключам словаря) и напечатаем книги, которые каждый из них написал:

In [ ]:
for author in books:            # Запомнив каждого автора в переменную author, 
                                # просмотрим книжную полку books и сделаем следующее:
    print(author, 'написал:')
    for book in books[author]:  # Запомнив каждую книгу в переменную book, 
                                # просмотрим список книг, написанных автором 
                                # author, и сделаем следующее:
        print(book)
    print()                     # Сделаем отступ между двумя авторами

Норберт Винер написал:
Кибернетика или управление и связь в животном и машине

Лев Выготский написал:
Мышление и речь
Психология искусства



Добавим возможность добавлять новые книги (и новых авторов), для этого:
1. пока человек не введёт слово "конец" будем спрашивать имя нового автора
2. если книг этого автора в библиотеке ещё нет:
    * создадим ключ с именем этого автора в нашем словаре книг, а в качестве списка книг укажем пустой список;
3. спросим название новой книги;
4. добавим новое название к списку книг автора.

Вывод списка авторов и книг скопируем из предыдущей ячейки.

In [ ]:
while True:
    new_author = input('Укажите автора: ')
    if new_author.lower() == 'конец':
        break
    if new_author not in books:   # Это новый автор, для него надо подготовить внутренний список
        books[new_author] = []    # Пока пусть побудет пустым
    
    # Мы закончили создавать пустой список для нового автора и дальше в любом случае спросим о новой книге
    new_book = input(f'Укажите название новой книги ({new_author}): ')  # Выведем строчку-подсказку, 
                                                                        # указав в ней имя автора
    books[new_author].append(new_book)  # Добавим к внутреннему списку книг автора new_author новую книгу

print()

for author in books:
    print(author, 'написал(написала):')
    for book in books[author]:
        print(book)
    print()

Укажите автора: Ася Казанцева
Укажите название новой книги (Ася Казанцева): В интернете кто-то неправ!
Укажите автора: конец

Норберт Винер написал(написала):
Кибернетика или управление и связь в животном и машине

Лев Выготский написал(написала):
Мышление и речь
Психология искусства

Ася Казанцева написал(написала):
В интернете кто-то неправ!



Итоговая программа этого раздела:

In [ ]:
# Создадим словарь, где ключом будет автор, а в качестве значений будет храниться список названий книг
books = {   
    'Норберт Винер': [  # Создадим список и наполним 1 книгой
        'Кибернетика или управление и связь в животном и машине'
    ],
    'Лев Выготский': [  # Создадим список и наполним 2 книгами
        'Мышление и речь',
        'Психология искусства'
    ]
}

while True:
    new_author = input('Укажите автора: ')
    if new_author.lower() == 'конец':
        break
    if new_author not in books:   # Это новый автор, для него надо подготовить внутренний список
        books[new_author] = []    # Пока пусть побудет пустым
    
    # Мы закончили создавать пустой список для нового автора и дальше в любом случае спросим о новой книге
    new_book = input(f'Укажите название новой книги ({new_author}): ')  # Выведем строчку-подсказку, 
                                                                        # указав в ней имя автора
    books[new_author].append(new_book)  # Добавим к внутреннему списку книг автора new_author новую книгу

print() # Добавим отступ между вводом новых книг (и авторов) и финальным выводом

for author in books:            # Запомнив каждого автора в переменную author, просмотрим книжную полку books 
                                # и сделаем следующее:
    print(author, 'написал:')
    
    for book in books[author]:  # Запомнив каждую книгу в переменную book, просмотрим список книг, написанных 
                                # авторов author, и сделаем следующее:
        print(book)
    print()                     # Сделаем отступ между двумя авторами

# Сортирока, минимум и максимум последовательностей

Когда мы работаем с данными, нам очень часто хочется данные каким-то образом отсортировать или найти минимальное или максимальное значение. 

Например, список студентов группы хочется напечатать по алфавиту. А словарь, в котором мы храним информацию о тратах по разным категориям, отсортировать по убыванию значений, чтобы увидеть, куда опять ушли деньги.

В Python есть функция `sorted()`, которая применяется к последовательностям и другим структурам данных и возвращает *отсортированный список* (даже если сортировался и не список).

Целые и вещественные числа `sorted()` отсортирует в порядке возрастания, а если в функции прописать параметр `reverse=True`, то порядок изменится на убывающий.

## Сортировка списка чисел

Начнем разбираться с функцией `sorted()` с самого простого примера —  отсортируем список, состощий из чисел.

In [ ]:
marks = [8.4, 4, 8, 9.4]
print(sorted(marks))                # Отсортировали список оценок по возрастанию
print(sorted(marks, reverse=True))  # А теперь по убыванию

[4, 8, 8.4, 9.4]
[9.4, 8.4, 8, 4]
[8.4, 4, 8, 9.4]


**ВАЖНО!** Функция `sorted()` не меняет список (или другую структуру данных), а создает новый.

In [ ]:
print(marks)  # Оригинальный список не изменился

[8.4, 4, 8, 9.4]


`sorted()` может отсортировать не только числа, но и другие типы данных. Главное, чтобы эти данные были **однотипные**.

In [ ]:
# Попытались отсортировать список из строк и целых чисел — получили ошибку.
print(sorted(['Ольга Ларина', 2003, 'Политология']))  

TypeError: ignored

## Сортировка списка строк и параметр key

Строки функция `sorted()` сортирует, можно сказать, по алфавиту. "Можно сказать", потому что алфавит, как его видит Python, отличается от нашего. Чтобы отсортировать строки Python подглядывает в специальную таблицу, в которой, например, сначала идут все цифры и какие-то специальные символы, потом все заглавные буквы, потом все строчные буквы... Давайте посмотрим на примере.


In [ ]:
fruits = ['apple', 'Яблоко', 'Апельсин', 'Orange', 'манго']
print(sorted(fruits))  # сортируем строки по возрастанию

['Orange', 'apple', 'Апельсин', 'Яблоко', 'манго']


Как мы видим, сначала идут заглавные латинские буквы, затем строчные латинские буквы, затем заглавные кириллические буквы и, наконец, строчные кириллические буквы. Наверное, когда мы думали о сортировке "по алфавиту", мы не рассчитывали, что регистр будет играть роль. Один вариант помочь Python корректно, с нашей точки зрения, отсортировать слова: привести все строки к нижнему регистру перед сортировкой.

In [ ]:
# Приводим все строки в нашим списке к нижему регистру с помощью функции map() и метода строк str.lower()
# Превращаем результат работы функции map в список
fruits_lower = list(map(str.lower, ['apple', 'Яблоко', 'Апельсин', 'Orange', 'манго']))

# Теперь работаем со списком строк в нижнем регистре
print(fruits_lower)  
print(sorted(fruits_lower))

['apple', 'яблоко', 'апельсин', 'orange', 'манго']
['apple', 'orange', 'апельсин', 'манго', 'яблоко']


Теперь алфавитный порядок для английских и русских слов сохранен, но мы потеряли оригинальное написание слов. Если нам важны заглавные буквы и мы не хотим от них избавляться, есть второй способ отсортировать строки без учета регистра. В этом нам поможет параметр функции `sorted()` — `key`. 

In [ ]:
fruits = ['apple', 'Яблоко', 'Апельсин', 'Orange', 'манго']
print(sorted(fruits, key=str.lower)) 

# 'Orange' теперь не идет раньше 'apple'
# Слова действительно отсортировались в алфавитном порядке, а регистр сохранился

['apple', 'Orange', 'Апельсин', 'манго', 'Яблоко']


Давайте разбираться. Параметру `key` внутри функции `sorted()` может указать название команды, которая потом применится ко всем элементам последовательности. Кажется, пока очень похоже на то, что делает `map()`. Но разница в том, что в этом случае команда применяется только на этапе сортировки, а исходная последовательность не изменяется. 

Давайте посмотрим, как этот алгоритм работает пошагово:

  * Функции `sorted()` попадает наш список фруктов. 
  * Параметру `key` присваивается метод строк `str.lower()`
  * Функция из параметра `key` применяется ко всем строкам в нашем списке, но не изменяет его, а *создает копию списка*, где все строки в нижнем регистре
  * Функция `sorted()` запоминает, какие строки из оригинального списка и его копии в нижнем регистре соответсвуют друг другу (например, 'Orange' -> 'orange').
  * Функция сортирует копию нашего списка, а потом расставляет элементы оригинального списка в нужные места (например, 'orange' на втором месте отсортированного списка заменяется обратно на 'Orange').

Параметр `key` очень полезный и когда вы научитесь писать собственные функции, вы сможете с помощью этого параметра делать многоуровневые сортировки сложных структур данных. Но начальном уровне его главное применение — сортировка строк в "алфавитном" порядке.

Еще одна неожиданность, которая может вас подстерегать при сортировке строк — это *сортировка чисел-строк*. Предположим, мы считали список чисел с клавиатуры и забыли их перевести в числовой формат перед сортировкой.

In [ ]:
exam_dates = input('Введите даты семинаров в октябре: ').split()
print(sorted(exam_dates)) # Отсортировали числа в формате str

Введите даты семинаров в октябре: 8 10 14 21 22
['10', '14', '21', '22', '8']


Как видим, даты отсортировались странновато. Но с точки зрения компьютера все в порядке: ведь, когда он видит строки, он идет смотреть в свою специальную таблицу, как их отсортировать, где символ "7" идет позже "1" и "2", а значит должен стоять позже них в отстортированном списке. Если первые символы совпадают, то сравниваются вторые, третьи... Точно также как мы бы с вами решали задачу такой сортировки "по алфавиту": мы ставим слово "яма" после "яблоко", потому что "м" идет позже по алфавиту, чем "б".

Python здесь не сравнивает числа между собой, а расставляет их в том порядке, в котором по его внутренней логике должны стоять символы, обозначающие цифры. Поэтому не забывайте переводить числа-строки в `int` или `float` перед сортировкой или хотя бы использовать `key=int` или `=float`, если не нужно дальше работать с ними как с числами.

In [ ]:
# Превратили строки внутри списка в int и отсортировали список чисел
print(sorted(map(int, exam_dates))) 

[8, 10, 14, 21, 22]


In [ ]:
# Отсортировали строки "как числа" с помощью key
sorted(exam_dates, key=int) 

['8', '10', '14', '21', '22']

## Нахождение минимума и максимума

Похожим образом со строками работают `min()` и `max()`, мы уже говорили немного про них, когда искали минимум и максимум в списках чисел. Посмотрим, как эти функции работают со списками чисел и строк.

In [ ]:
marks = [8.4, 4, 8, 9.4]

# Нашли минимальный и максимальный элемент в списке из чисел
print(f'Минимальная оценка: {min(marks)}, максимальная оценка: {max(marks)}') 

names = ['Аня', 'Алена', 'Валя', 'Дима', 'Таня', 'Катя']
# Нашли строки, которые идут раньше всех и позже всех по алфавиту по мнению компьютера 
# (в этом примере оно совпадает с нашим)
print(f'Раньше все по алфавиту идет имя "{min(names)}", а позже всех — "{max(names)}"') 

Минимальная оценка: 4, максимальная оценка: 9.4
Раньше все по алфавиту идет имя "Алена", а позже всех — "Таня"


Мы с вами разобрали сортировки списков, но не забывайте, что `sorted()`, `min()` и `max()` можно использовать и с другими типами данных — строками, кортежами, множествами. Эти функции с ними будут работать почти так же как со списками. В результате сортировки кортежей и множеств мы получим список их отсортированных элементов, а в случае строки — список ее отсортированных символов. 

Со словарями сортировка тоже будет работать, но из-за того, что информация в словаре хранится парами, нам придется изобрести некоторые хитрости, чтобы их сортировать по значениям, поэтому о них мы поговорим отдельно.


# Подсчёт статистики с помощью словарей
С помощью словарей удобно собирать статистику. Например, когда мы хотим посчитать собственные расходы и понять, на что мы тратим деньги.
После каждой траты денег мы будем записывать в блокнот строчку вида "`<категория траты> <сумма>`".

Подсчитаем статистику, для этого:
1. создадим пустой словарь для хранения данных о наших тратах, где ключом будет строка с названием категории, а значением — сумма трат;
2. пока нам не напишут слово "конец":
    1. будем спрашивать строку с категорией и суммой;
    2. будем разбивать полученную строку на список из названия категории и суммы трат;
    3. будем сохранять информацию в словарь.


In [ ]:
stat = {} # Словарь для хранения статистики

while True:
    info = input('Укажите категорию и, через пробел, сумму: ') # Спросим категорию трат и сумму одной строкой
    if info.lower() == 'конец':   # Если нам дали слово "конец" вместо категории,
        break                     # закончим подсчёт статистики

    data = info.split()           # Разделим введённую строку на список,
    cat = data[0]                 # из которого заберём категорию трат
    amount = int(data[1])         # и сумму
    stat[cat] += amount

print(stat)

Укажите категорию и, через пробел, сумму: Аренда 15000


KeyError: ignored

Разберём ошибку выше: в строке `stat[cat] += amount` компьютер ещё ничего не знает про такие расходы, как "Аренда". Научим его проверять, нужно ли создать новую категорию, с помощью уже знакомой нам конструкции `if .. in`:


In [ ]:
stat = {'Аренда': 15000}               # Укажем, что за аренду мы уже заплатили 15000
cat = input('Укажите категорию: ')
if cat not in stat:
    print('Такой категории ещё нет!')
    stat[cat] = 0                      # Создадим для категории ключ и укажем, что пока трат не было
else:                                  # Если категория уже есть
    print('Такая категория уже есть!')
    
print(f'{cat}: потрачено {stat[cat]} рублей.') 

Укажите категорию: Еда
Такой категории ещё нет!
Еда: потрачено 0 рублей.


Добавим проверку существования категории в наш код:

In [ ]:
stat = {}

while True:
    info = input('Укажите категорию и, через пробел, сумму: ')
    if info.lower() == 'конец':
        break 

    data = info.split()
    cat = data[0]
    amount = int(data[1])

    if cat not in stat:         # Если мы обрабатываем такую категорию впервые:
        stat[cat] = amount      # добавим новую категорию
    else:                       # Если мы уже знаем про такую категорию:
        stat[cat] += amount     # увеличим сумму трат

print(stat)

Укажите категорию и, через пробел, сумму: Аренда 15000
Укажите категорию и, через пробел, сумму: Транспорт 3500
Укажите категорию и, через пробел, сумму: конец
{'Аренда': 15000, 'Транспорт': 3500}


Добавим красивый вывод. Так как для нас важна и категория трат, и сумма трат по этой категории, попросим компьютер перебирать сразу и ключ, и значение словаря. Сделать это можно с помощью метода `.items()`. Посмотрим, как его видит Python:

In [ ]:
print(type(stat.items()))

<class 'dict_items'>


Посмотрим, что лежит внутри:

In [ ]:
print(stat.items())

dict_items([('Аренда', 15000), ('Транспорт', 3500)])


Мы видим, что это список кортежей, где под 0 индексом всегда идёт ключ, а под 1 индексом — соответствующее ему значение. Мы можем перебрать такой список кортежей с помощью цикла `for`:

In [ ]:
for info in stat.items():
    print(f'По категории "{info[0]}" были траты на {info[1]} рублей.')

По категории "Аренда" были траты на 15000 рублей.
По категории "Транспорт" были траты на 3500 рублей.


Однако мы можем извлечь из кортежа информацию сразу в переменные с помощью конструкции `переменная, другая_переменная = (первый элемент, второй элемент)`. В этой конструкции в первую переменную будет сохранено значение из первого элемента кортежа, а во вторую переменную —  значение из второго элемента:

In [ ]:
cat, amount = ('Аренда', 15000)
print(f'По категории {cat} были траты на {amount} рублей.')

По категории Аренда были траты на 15000 рублей.


Воспользуемся конструкцией `for переменная,другая_переменная in stat.items()` для перебора всех категорий трат:

In [ ]:
for cat, amount in stat.items():  
    print(f'По категории {cat} были траты на {amount} рублей.')

По категории Аренда были траты на 15000 рублей.
По категории Транспорт были траты на 3500 рублей.


В предыдущем примере ключи словаря, которыми выступают категории трат, сохраняются в переменную ``cat``, а значения, которыми являются суммы трат, сохраняются в переменную ``amount``.

Очень часто мы тратим много денег не на категории с самыми дорогими тратами, а на категории с самым большим количеством маленьких трат. Например, когда постоянно покупаем дешёвые шоколадки. 
Научим нашу программу запоминать не только сумму трат по каждой категории, но и детализацию трат по датам покупок, для чего поменяем наш словарь со статистикой, превратив его в *словарь списков списков*:

* Словарь будет содержать категории трат и список трат
    * каждый список трат будет содержать информацию о тратах
        * каждая трата будет содержать сумму траты и дату, когда была совершена покупка.

но можно и хранить для каждой категории список, где на 0 месте будем хранить количество, а на 1 месте — сумму трат по категории:

In [ ]:
stat = {}

while True:
    info = input('Укажите категорию, сумму и дату траты: ')
    if info.lower() == 'конец':
        break 

    data = info.split()
    cat = data[0]
    amount = int(data[1])
    date = data[2] 

    if cat not in stat:              # Если мы обрабатываем такую категорию впервые:
        stat[cat] = []               # Подготовим наш список списков
    stat[cat].append([amount, date]) # Добавим в список списков трат данной категории список с суммой и датой
    

for cat in stat:
    print(f'По категории {cat} было {len(stat[cat])} трат:')
    for amount, date in stat[cat]:
        print(f'{date}: {amount} рублей.')
    

# Выведем на экран структуру получившегося словаря:
print('stat =', stat)

Укажите категорию, сумму и дату траты: Вкусняшки 150 2020-09-02
Укажите категорию, сумму и дату траты: Вкусняшки 270 2020-09-06
Укажите категорию, сумму и дату траты: Аренда 2700 2020-09-05
Укажите категорию, сумму и дату траты: Аренда 25000 2020-09-05
Укажите категорию, сумму и дату траты: Транспорт 72 2020-09-01
Укажите категорию, сумму и дату траты: конец
По категории Вкусняшки было 2 трат:
2020-09-02: 150 рублей.
2020-09-06: 270 рублей.
По категории Аренда было 2 трат:
2020-09-05: 2700 рублей.
2020-09-05: 25000 рублей.
По категории Транспорт было 1 трат:
2020-09-01: 72 рублей.
stat = {'Вкусняшки': [[150, '2020-09-02'], [270, '2020-09-06']], 'Аренда': [[2700, '2020-09-05'], [25000, '2020-09-05']], 'Транспорт': [[72, '2020-09-01']]}


# Сортирока, минимум и максимум словарей
Нам понравилось отслеживать свои расходы и теперь мы всегда вносим их в программу.

Нас интересует, как мы тратим деньги: на что потратили больше всего, в какие дни мы тратили больше всего, интересует упорядоченная таблица категорий трат.

Посмотрим ещё раз на наш *словарь списков списков*:

In [ ]:
# Словарь вида категория: список списков трат
expenses = {        
    'Вкусняшки': [            # Список списков трат
        [150, '2020-09-02'],  # Каждая трата — это пара из суммы и даты
        [270, '2020-09-06'] 
    ], 
    'Аренда': [
        [2700, '2020-09-05'], 
        [25000, '2020-09-05']
    ], 
    'Транспорт': [[72, '2020-09-01']]   # В этом списке списков трат только одна трата
}

Мы перенесли все наши расходы за неделю в Python:

In [ ]:
expenses = {
    'Вкусняшки': [
        [150, '2020-09-02'],
        [270, '2020-09-04'],
        [270, '2020-09-06']
    ],
    'Продукты': [
        [2424.8, '2020-09-02'],
        [1524.65, '2020-09-06']
    ],
    'Кафе': [
        [790, '2020-09-01'],
        [524, '2020-09-03'],
        [524, '2020-09-07']
    ],
    'Транспорт': [
        [144, '2020-09-01'],
        [72, '2020-09-02'],
        [474, '2020-09-04']
    ],
    'Аренда': [
        [25000, '2020-09-05'],
        [2743.3, '2020-09-05']
    ]
}

Нам удобно хранить так наши траты, но не очень удобро обрабатывать. Посчитаем отдельно, сколько мы потратили по каждой категории и в каждый день, для этого:

1. Создадим два пустых словаря для хранения статистики "всего по категории" и "всего по дате".
2. Пройдём по каждой категории словаря и посчитаем сумму трат в списке списков трат по категории.
3. Пройдём по всем тратам во всех категориях и просуммируем траты в каждый из дней.

In [ ]:
ex_dates = {}
ex_cat = {}

for cat in expenses:
    if cat not in ex_cat:       # Если мы ещё не считали траты в этой категории:
        ex_cat[cat] = 0         # Подготовим для себя ключ и значение в словаре трат по категориям
    for exp in expenses[cat]:
        ex_cat[cat] += exp[0]   # Добавим сумму трат из 0 элемента списка "Сумма, дата"

for cat in expenses:
    for exp in expenses[cat]:
        date = exp[1]
        if date not in ex_dates:
            ex_dates[date] = 0
        ex_dates[date] += exp[0]

print(ex_cat)
print(ex_dates)

{'Вкусняшки': 690, 'Продукты': 3949.4500000000003, 'Кафе': 1838, 'Транспорт': 690, 'Аренда': 27743.3}
{'2020-09-02': 2646.8, '2020-09-04': 744, '2020-09-06': 1794.65, '2020-09-01': 934, '2020-09-03': 524, '2020-09-07': 524, '2020-09-05': 27743.3}


Чудесно. Теперь у нас есть траты за каждый день и по каждой категории, давайте узнаем, как в них искать минимум и максимум, как их сортировать.

Из-за структуры, в которой элементы хранятся парами ключ-значение, сортировка словаря не такая простая, как сортировка списков и других структур данных.

Давайте попробуем отсортировать словарь, где ключами являются целые числа:

In [ ]:
# Словарь, в котором ключ оценка-число, а значение — категория, к которой она относится
marks_to_words = {5:'хорошо', 4:'удовлетворительно', 8:'отлично', 10:'отлично'}

print(sorted(marks_to_words))  # Попытались отсортировать словарь — получили отсортированный список ключей,
print(marks_to_words.keys())   # но сам объект, храняющий ключи не отсортировался

[4, 5, 8, 10]
dict_keys([5, 4, 8, 10])


Получили отсортированный список ключей, но сам словарь не изменился. Если мы хотим вывести значения словаря или пары ключ-значения, например, в порядке возрастания ключей, то мы можем использовать отсортированный список ключей в цикле `for`.

Ключи в словаре уникальны и каждый соответствует своему значению. А значит не важно, в каком порядке мы эти ключи будем перебирать.

In [ ]:
for key in sorted(marks_to_words):
    print(f'Оценке {key} соответсвует {marks_to_words[key]}')

Оценке 4 соответсвует удовлетворительно
Оценке 5 соответсвует хорошо
Оценке 8 соответсвует отлично
Оценке 10 соответсвует отлично


`min()` и `max()` тоже вернут нам минимальное и максимальное значение для ключей словаря:

In [ ]:
print(f'Самая низкая оценка: {min(marks_to_words)}')
print(f'Самые высокая оценка: {max(marks_to_words)}')

Самая низкая оценка: 4
Самые высокая оценка: 10


`min()` и `max()` могут найти нам самую маленькую и самую большую строку. Например, если мы хотим найти первый и последний день наших наблюдений за расходами:

In [ ]:
print(f'Первая дата: {min(ex_dates)}')
print(f'Последняя дата: {max(ex_dates)}')

Первая дата: 2020-09-01
Последняя дата: 2020-09-07


Так как мы написали ведущие нули у дат, то строки будут корректно сортироваться, и '2020-09-01' будет идти до '2020-09-10'.

Выведем наши расходы по датам по порядку: от первой даты наблюдений к последней:

In [ ]:
for date in sorted(ex_dates.keys()):
    print(f'{date}: потрачено {ex_dates[date]} рублей.')

2020-09-01: потрачено 934 рублей.
2020-09-02: потрачено 2646.8 рублей.
2020-09-03: потрачено 524 рублей.
2020-09-04: потрачено 744 рублей.
2020-09-05: потрачено 27743.3 рублей.
2020-09-06: потрачено 1794.65 рублей.
2020-09-07: потрачено 524 рублей.


Напечатали даты в хронологическом порядке. Неплохо, но, наверное, нам все-таки опять интереснее узнать, когда же мы потратили больше денег или в какой категории. Но эта информация теперь в значениях. Как быть?

Мы с вами говорили про метод словаря `.values()`, который дает нам некий объект, в котором содержатся значения. Давайте попробуем в нашем словарике трат по категориям его отсортировать:

In [ ]:
# Объект, в котором содержатся значения нашего словаря
print(ex_cat.values())

# Отсортированный список значений нашего словаря
print(sorted(ex_cat.values()))  

dict_values([690, 3949.4500000000003, 1838, 690, 27743.3])
[690, 690, 1838, 3949.4500000000003, 27743.3]


Получилось! Но сложность в том, что обращаться по ключу к значению мы умеем, а наоборот — нет. И на самом деле простого способа сделать это встроенными средствами Python — нет.

Давайте попробуем схитрить: напишем цикл `for`, который для каждого значения в отсортированном списке будет перебирать все ключи словаря и проверять, не равно ли значение конкретного ключа этому значению в отсортированном списке.

In [ ]:
for value in sorted(ex_cat.values()):
    print(f'Ищем категорию, в которой потратили {value} рублей:')
    for cat in ex_cat:
        if ex_cat[cat] == value:
            print(cat, ex_cat[cat])

Ищем категорию, в которой потратили 690 рублей:
Вкусняшки 690
Транспорт 690
Ищем категорию, в которой потратили 690 рублей:
Вкусняшки 690
Транспорт 690
Ищем категорию, в которой потратили 1838 рублей:
Кафе 1838
Ищем категорию, в которой потратили 3949.4500000000003 рублей:
Продукты 3949.4500000000003
Ищем категорию, в которой потратили 27743.3 рублей:
Аренда 27743.3


Почти получилось. Единственная проблема — категории с одинаковыми расходами напечатались по два раза, потому что значения в отличие от ключей могут быть не уникальны, и мы два раза нашли категории, в которых мы потратили 690 рублей. Давайте попробуем решить это проблему, превратив объект со значениями словаря в множество перед сортировкой:

In [ ]:
print(f'Все значения словаря: {ex_cat.values()}')
print(f'В нашем словаре {len(ex_cat.values())} значений')

print(f'Множество уникальных значений словаря: {set(ex_cat.values())}')
print(f'Но из них только {len(set(ex_cat.values()))} уникальных')

Все значения словаря: dict_values([690, 3949.4500000000003, 1838, 690, 27743.3])
В нашем словаре 5 значений
Множество уникальных значений словаря: {690, 3949.4500000000003, 1838, 27743.3}
Но из них только 4 уникальных


In [ ]:
for value in sorted(set(ex_cat.values())):
    print(f'Ищем категорию, в которой потратили {value} рублей:')
    for cat in ex_cat:
        if ex_cat[cat] == value:
            print(cat, ex_cat[cat])

Ищем категорию, в которой потратили 690 рублей:
Вкусняшки 690
Транспорт 690
Ищем категорию, в которой потратили 1838 рублей:
Кафе 1838
Ищем категорию, в которой потратили 3949.4500000000003 рублей:
Продукты 3949.4500000000003
Ищем категорию, в которой потратили 27743.3 рублей:
Аренда 27743.3


Вот теперь точно получилось! Нужно заметить, что такой код — не единственное решение такой задачи (и если честно не самый оптимальный, потому что при таком переборе мы делаем много лишних сравнений). Позже в курсе мы покажем еще один вариант того, как можно отсортировать словарь по значению.


С сортировкой разобрались, теперь давайте попробуем найти в наших недельных тратах категорию с самыми дольшими расходыми (или несколько!). И это уже попроще. Нам всего лишь нужно найти минимальное и максимальное значение в словаре, а потом в цикле найти ключи, которые этим значениям соответствуют.

In [ ]:
max_exp = max(ex_cat.values())
min_exp = min(ex_cat.values())

for key in ex_cat:
    if ex_cat[key] == max_exp:
        print(f'Категория с самыми большими тратами —  {key}')
    elif ex_cat[key] == min_exp:
        print(f'Категория с самыми маленькими тратами —  {key}')

Категория с самыми маленькими тратами —  Вкусняшки
Категория с самыми маленькими тратами —  Транспорт
Категория с самыми большими тратами —  Аренда


Категорий с самыми маленькими тратами две, но это не ошибка. В нашем словаре `expenses` две категории действительно имеют одинаковое и минимальное значение расходов.

Мы с вами разобрали сортировки списков и словарей, но не забывайте, что `sorted()`, `min()` и `max()` можно использовать и с другими типами данных — строками, кортежами, множествами. Так как в этих структурах нет таких особенностей, как в словарях, где элементы представляются собой пары ключ-значения, то функции с ними будут работать почти так же как со списками. В результате сортировки кортежей и множеств мы получим список их отсортированных элементов, а в случае строки — список ее отсортированных символов. 

# Словари словарей
Иногда информация становится настолько сложной, что нам не хватает просто словарей или словарей списков. Что мы можем знать о каждой книге?
*   название,
*   год выпуска,
*   издательство,
*   ...

Доработаем наш пример о библиотеке, но теперь о каждом авторе будем хранить не просто название его книги, а много информации. Поможет нам с этим **словарь словарей**. Для простоты мы пока будем запоминать информацию только об одной книге для каждого автора.



Создадим словарь словарей ``books``:
* ключом будет строка из имени и фамилии автора,
* значением будeт словарь информации о книге, у которого будут внутренние ключи и значения:
    * title: название книги,
    * year: год издания,
    * publisher: издательство.

In [ ]:
books = {
    'Норберт Винер': {
        'title' : 'Кибернетика или управление и связь в животном и машине',
        'year'  : 1958, 
        'publisher' : 'Советское радио'
    },
    'Лев Выготский': {
        'title': 'Мышление и речь',
        'year' : 1999,
        'publisher' : 'Лабиринт'
    }
}
print(books)

{'Норберт Винер': {'title': 'Кибернетика или управление и связь в животном и машине', 'year': 1958, 'publisher': 'Советское радио'}, 'Лев Выготский': {'title': 'Мышление и речь', 'year': 1999, 'publisher': 'Лабиринт'}}


Мы можем вывести информацию о книге, обратившись к ней через имя автора:

In [ ]:
print(books['Норберт Винер'])
print(books['Лев Выготский'])

{'title': 'Кибернетика или управление и связь в животном и машине', 'year': 1958, 'publisher': 'Советское радио'}
{'title': 'Мышление и речь', 'year': 1999, 'publisher': 'Лабиринт'}


Наконец, мы можем вывести конкретную информацию о книге (название, например), если обратимся к внутреннему словарю по *имени автора* и к нужному нам ключу во внутреннем словаре по слову *title*:

In [ ]:
print(books['Норберт Винер']['title'])
print(books['Лев Выготский']['title'])

Кибернетика или управление и связь в животном и машине
Мышление и речь


Мы можем перебрать такой словарь с помощью цикла `for author, book in books.items()`: будем сохранять имя и фамилию автора книги из ключа нашего словаря ``books`` в переменную ``author``, а словарь с информацией о книге — в словарь ``book``:

In [ ]:
books = {
    'Норберт Винер': {
        'title' : 'Кибернетика или управление и связь в животном и машине',
        'year'  : 1958, 
        'publisher' : 'Советское радио'
    },
    'Лев Выготский': {
        'title': 'Мышление и речь',
        'year' : 1999,
        'publisher' : 'Лабиринт'
    }
}

for author, book in books.items():
    print(f'{author}. {book["title"]} — "{book["publisher"]}", {book["year"]}г.')

Норберт Винер. Кибернетика или управление и связь в животном и машине — "Советское радио", 1958г.
Лев Выготский. Мышление и речь — "Лабиринт", 1999г.


Доработаем пример, чтобы можно было указывать новых авторов, для чего:
1. создадим словарь словарей ``books``, где ключом будет выступать строка из имени и фамилии автора, а значением — словарь информации о книге,
2. пока нам не напишут слово "конец", будем спрашивать информацию о новых авторах и их книгах:
    1. если этот автор уже внесён в словарь:
        * будем говорить, что мы пока не умеем редактирвоать информацию об имеющихся авторах
    2. если такого автора ещё нет:
        * создадим новый внутренний словарь

In [ ]:
books = {
    'Норберт Винер': {
        'title' : 'Кибернетика или управление и связь в животном и машине',
        'year'  : 1958, 
        'publisher' : 'Советское радио'
    },
    'Лев Выготский': {
        'title': 'Мышление и речь',
        'year' : 1999,
        'publisher' : 'Лабиринт'
    }
}

while True:
    new_author = input('Укажите нового автора: ')
    if new_author.lower() == 'конец':
        break
    if new_author in books:             # Сейчас мы не делаем часть с обновлением информации
        print('Такой автор уже есть!')  # о существующих авторах
    else:
        books[new_author] = {}  # Создадим пустой словарь для хранения 
                                # информации о новом авторе
        
        # Спросим название книги и положим в этот внутренний словарь с ключом title
        books[new_author]['title'] = input('Как называется книга? ')
        
        # Спросим год издания книги и положим во внутренний словарь с ключом year
        books[new_author]['year'] = int(input('В каком году выпущена? '))
        
        # Спросим издательство, выпустившее книгу, и положим во внутренний словарь
        # с ключом publisher
        books[new_author]['publisher'] = input('Каким издательством? ') 

for author, book in books.items():
    print(f'{author}. {book["title"]} — "{book["publisher"]}", {book["year"]}г.')

Укажите нового автора: Лев Выготский
Такой автор уже есть!
Укажите нового автора: Дильшат Харман
Как называется книга? Страдающее Средневековье. Парадоксы христианской иконографии
В каком году выпущена? 2018
Каким издательством? АСТ
Укажите нового автора: конец
Норберт Винер. Кибернетика или управление и связь в животном и машине — "Советское радио", 1958г.
Лев Выготский. Мышление и речь — "Лабиринт", 1999г.
Дильшат Харман. Страдающее Средневековье. Парадоксы христианской иконографии — "АСТ", 2018г.


Мы могли бы не создавать отдельно пустой внутренний словарь, а сразу спрашивать значения в процессе создания словаря. Посмотрим на простом примере словаря с информацией о книге Норберта Винера, как это работает:

In [ ]:
wiener = {
    'title': input('Как называется книга? '),
    'year': int(input('В каком году выпущена? ')),
    'publisher': input('Каким издательством? ')
}
print(f'Информация о книге Норберта Винера: {wiener}.')

Как называется книга? Кибернетика или управление и связь в животном и машине
В каком году выпущена? 1958
Каким издательством? Советское радио
Информация о книге Норберта Винера: {'title': 'Кибернетика или управление и связь в животном и машине', 'year': 1958, 'publisher': 'Советское радио'}.


Доработаем программу для библиотеки:

In [ ]:
books = {
    'Норберт Винер': {
        'title' : 'Кибернетика или управление и связь в животном и машине',
        'year'  : 1958, 
        'publisher' : 'Советское радио'
    },
    'Лев Выготский': {
        'title': 'Мышление и речь',
        'year' : 1999,
        'publisher' : 'Лабиринт'
    }
}

while True:
    new_author = input('Укажите нового автора: ')
    if new_author.lower() == 'конец':
        break
    if new_author in books:             # Сейчас мы не делаем часть с обновлением информации
        print('Такой автор уже есть!')  # о существующих авторах
    else:
        books[new_author] = {           # Создадим новый словарь для хранения информации
                                        # о книге и сразу спросим:
            'title': input('Как называется книга? '),       # название,
            'year': int(input('В каком году выпущена? ')),  # год издания,
            'publisher': input('Каким издательством? ')     # издательство
        }   

for author, book in books.items():
    print(f'{author}. {book["title"]} — "{book["publisher"]}", {book["year"]}г.')

Укажите нового автора: Лев Выготский
Такой автор уже есть!
Укажите нового автора: Дильшат Харман
Как называется книга? Страдающее Средневековье. Парадоксы христианской иконографии
В каком году выпущена? 2018
Каким издательством? АСТ
Укажите нового автора: конец
Норберт Винер. Кибернетика или управление и связь в животном и машине — "Советское радио", 1958г.
Лев Выготский. Мышление и речь — "Лабиринт", 1999г.
Дильшат Харман. Страдающее Средневековье. Парадоксы христианской иконографии — "АСТ", 2018г.


Предположим, мы хотим организовать возможность изменять информацию о книге уже имеющегося в картотеке автора. В этом случае мы можем просто удалять информацию о книгах имеющихся авторов с помощью функции `del()`, а потом заново создавать внутренний словарь с информацией о книге. Посмотрим, как работает функция `del()`:

In [ ]:
vygotsky = {
    'title': 'Мышление и речь',
    'year' : 1999,
    'publisher' : 'Лабиринт'
}
print(f'Информация о книге Льва Выготского: {vygotsky}.')
del(vygotsky['year'])     # Удалим год издания
print(f'Без года издания: {vygotsky}.')

Информация о книге Льва Выготского: {'title': 'Мышление и речь', 'year': 1999, 'publisher': 'Лабиринт'}.
Без года издания: {'title': 'Мышление и речь', 'publisher': 'Лабиринт'}.


Если `del()` не сможет удалить ключ, она выдаст ошибку:

In [ ]:
vygotsky = {
    'title': 'Мышление и речь',
    'year' : 1999,
    'publisher' : 'Лабиринт'
}
print(f'Информация о книге Льва Выготского: {vygotsky}.')
del(vygotsky['year'])
del(vygotsky['year'])     # Попробуем удалить уже отсутствующий год издания
print(f'Без года издания: {vygotsky}.')

Информация о книге Льва Выготского: {'title': 'Мышление и речь', 'year': 1999, 'publisher': 'Лабиринт'}.


KeyError: ignored

Мы можем удалить даже внутренний словарь целиком:

In [ ]:
books = {
    'Норберт Винер': {
        'title' : 'Кибернетика или управление и связь в животном и машине',
        'year'  : 1958, 
        'publisher' : 'Советское радио'
    },
    'Лев Выготский': {
        'title': 'Мышление и речь',
        'year' : 1999,
        'publisher' : 'Лабиринт'
    }
}
print('Удалим всю информацию о книге Льва Выготского и выведем оставшийся словарь:')
del(books['Лев Выготский'])
print(books)

Удалим информацию о книге Льва Выготского и выведем оставшийся словарь.
{'Норберт Винер': {'title': 'Кибернетика или управление и связь в животном и машине', 'year': 1958, 'publisher': 'Советское радио'}}


и даже переменную саму по себе:

In [ ]:
year = 2077
print(f'Год: {year}')
del(year)
print(f'Год: {year}')   # В этой строке переменная year уже не существует

Год: 2077


NameError: ignored

Доработаем наш пример о библиотеке. Теперь, если человек укажет уже существующего автора, мы удалим информацию об имеющейся книге, а затем заново создадим и заполним словарь:

In [ ]:
books = {
    'Норберт Винер': {
        'title' : 'Кибернетика или управление и связь в животном и машине',
        'year'  : 1958, 
        'publisher' : 'Советское радио'
    },
    'Лев Выготский': {
        'title': 'Мышление и речь',
        'year' : 1999,
        'publisher' : 'Лабиринт'
    }
}

while True:
    new_author = input('Укажите нового автора: ')
    if new_author.lower() == 'конец':
        break
    if new_author in books:     # Если такой автор уже есть в словаре books:
        print('Такой автор уже есть! Удаляю...') 
        del books[new_author]   # В этой строке мы удаляем весь внутренний словарь
        print('Удалил информацию об авторе, укажите новую:')
    
    # Теперь мы в любом случае создаём новый внутренний словарь и заполняем его
    books[new_author] = {
        'title': input('Как называется книга? '),
        'year': int(input('В каком году выпущена? ')),
        'publisher': input('Каким издательством? ')
    }

print() # Добавим отступ перед выводом списка книг в библиотеке ради красоты и гармонии

for author, book in books.items():
    print(f'{author}. {book["title"]} — "{book["publisher"]}", {book["year"]}г.')

Укажите нового автора: Лев Выготский
Такой автор уже есть! Удаляю...
Удалил информацию об авторе, укажите новую:
Как называется книга? Психология искусства
В каком году выпущена? 1986
Каким издательством? Искусство
Укажите нового автора: конец

Норберт Винер. Кибернетика или управление и связь в животном и машине — "Советское радио", 1958г.
Лев Выготский. Психология искусства — "Искусство", 1986г.


Можем ли мы запомнить много книг для каждого автора, и про каждую из книг заполнить и год издания, и издательство, и количество страниц?

Да, можем.

Нам поможет в этом **словарь списков словарей**, но это, как говорится в одной известной телепрограмме, *уже совсем другая история*.